sdformat_urdf/README.md

```diff
@@ -0,0 +1,55 @@
+# SDFormat URDF
+
+This package contains a C++ library and `urdf_parser_plugin` for converting SDFormat XML into URDF C++ structures.
+Installing it allows one to use SDFormat XML instead of URDF XML as a robot description.
+
+## Supported platforms
+
+The package has been tested with ROS Rolling Ridley on Ubuntu Focal.
+
+## Limitations and Considerations
+
+In general, URDF C++ structures cannot represent every possible SDFormat XML file.
+SDFormat XML used with this plugin must be constructed with the following limitations.
+
+### Limitations that result in Errors
+
+This package will error and refuse to convert any SDFormat XML that violates these constraints.
+
+#### Model limitations
+* The XML file must contain a single `<model>` not in a `<world>`
+* The `<model>` tag must not have a `<pose>`
+* There must not be any nested `<model>`
+
+#### Geometry limitations
+* The only supported geometry types are `<box>`, `<cylinder>`, `<mesh>`, and `<sphere>`
+* The model must not use `<plane>` or `<heightmap>`
+
+#### Joint limitations
+* The only supported joint types are `continuous`, `fixed`, `prismatic`, and `revolute`
+* The model must not use `universal`, `screw`, `revolute2`, `gearbox`, or `ball` joints
+
+#### Kinematic limitations
+*  Starting from the canonical link, the links and joints must form a tree
+*  No link may be a child of more than 1 joint
+*  The canonical link cannot be a child of any joint
+
+### Limitations that may result in Warnings
+
+If any of these constraints are voilated then the library may issue a console warning, but the model is still converted to URDF c++ structures.
+The warning is issued using an `rcutils` logger with the name `sdformat_urdf`.
+
+#### Joint limitations
+* `<axis>` should not use `<initial_position>`
+* `<dynamics>` should not use `<spring_reference>` or `<spring_stiffness>`
+* `<limit>` should not use `<dissipation>` or `<stiffness>`
+* `<joint>` should not use `<sensor>` or `<physics>`
+
+#### Link limitations
+* `<link>` should not have any `<light>` or `<sensor>`
+
+#### Material limitations
+* Only solid color materials are supported - the color of the urdf comes from the `<ambient>` tag of the material
```

Comment:
IIUC ambient and diffuse are added together to make the final color. Ambient is always visible because it models light that bounced off so many surfaces that the source is unknown. Diffuse is only visible when a light is pointed at it. I went with ambient since it's always visible, but it might not be a great choice if the model was made for an environment with no ambient light, like space.

It could be possible to combine the two in some way, but adding a diffuse component means making an assumption about the color and intensity of the light shining on the object. If the object is modeled for a dark space with bright directional lights then it would make sense to add a lot of the diffuse (which means assume it's illuminated by a bright white light), but If the object was modeled for a bright space with lots of ambient light but dim directional lights then a bright white light might make the final color near-white.

http://gazebosim.org/tutorials?tut=color_model

Reply:
Okay. So in that model, I would say diffuse is closer to the "color" of the material because in real life, the color of objects does depend on the light shining on them. In Rviz, the default light is a white light, but [this](https://github.com/ros-visualization/rviz/pull/1146) PR suggests it's possible to add different types of lights. In that case, I would think users want the objects to be affected by the different types of lights available in the environment.

Also from looking at Rviz, it looks like the URDF color is set as both the diffuse and the ambient (with half the intensity). 
https://github.com/ros-visualization/rviz/blob/8bd56fb928b255e095424a5f1fb3ff8a5274602e/src/rviz/robot/robot_link.cpp#L473

Reply:
> I would say diffuse is closer to the "color" of the material because in real life, the color of objects does depend on the light shining on them

`<ambient>` is the same. The color depends on the ambient light color, but ambient light is a [property of the environment rather than a directional light](http://sdformat.org/spec?ver=1.7&elem=scene#scene_ambient). The odd bit about this is the `urdf` color is like the final color one would see when the model is rendered. `sdformat_urdf` using only `<ambient>` as the color is like rendering the model in the scene with `rgba(1 1 1 1)` ambient light and no directional lights. RViz is un-rendering that color to `color * 0.5 * ambient` and `color * diffuse` components and re-rendering it to the screen.

Since we need to render to a color, maybe we can use the default lighting setup in Gazebo as the example? The default gazebo empty world has [`rgba(0.4 0.4 0.4 1)` ambient light](http://sdformat.org/spec?ver=1.7&elem=scene#scene_ambient) with a directional light having [diffuse `rgba(0.8 0.8 0.8 1)`](https://github.com/ignitionrobotics/ign-gazebo/blob/da2ddbc14bc487d53e49de5b9fdea3272797698c/examples/worlds/default.sdf#L36) and [specular `rgba(0.2 0.2 0.2 1)`](https://github.com/ignitionrobotics/ign-gazebo/blob/da2ddbc14bc487d53e49de5b9fdea3272797698c/examples/worlds/default.sdf#L37). Specular tends to just be a highlight, so I think that could be ignored. That would leave `urdf color = 0.4 * sdf_ambient + 0.8 * sdf_diffuse`.

Reply:
Uses `0.4 * ambient + 0.8 * diffuse` to render color in c84405d

Reply:
Okay, sounds good to me.